parquet/src/arrow/arrow_reader/mod.rs

```diff
@@ -353,21 +354,31 @@ impl ParquetRecordBatchReader {
 pub(crate) fn evaluate_predicate(
     batch_size: usize,
     array_reader: Box<dyn ArrayReader>,
-    selection: Option<RowSelection>,
+    selection: &Option<RowSelection>,
     predicate: &mut dyn ArrowPredicate,
-) -> Result<RowSelection> {
+) -> Result<(RowSelection, RecordBatch)> {
     let reader =
         ParquetRecordBatchReader::new(batch_size, array_reader, selection.clone());
     let mut filters = vec![];
+    let mut batches = vec![];
+
+    let schema = reader.schema();
+
     for maybe_batch in reader {
-        filters.push(predicate.filter(maybe_batch?)?);
+        let batch = maybe_batch?;
+        let filter = predicate.filter(&batch)?;
+        batches.push(filter_record_batch(&batch, &filter)?);
+        filters.push(filter);
     }
 
-    let raw = RowSelection::from_filters(&filters);
-    Ok(match selection {
-        Some(selection) => selection.and(&raw),
-        None => raw,
-    })
+    let batch = RecordBatch::concat(&schema, &batches)?;
```

Comment:
If you have a not very selective predicate, won't this massively balloon your memory consumption? As you will effectively hydrate the entire row group into a RecordBatch?

Reply:
Perhaps we could have a hybrid approach where if the predicate isn't proving very selective, we bail out? :thinking:

Reply:
Not sure I follow. What would get pushed into the parquet reader?

Reply:
If we've buffered too much in memory, which is highly likely with any non-selective predicate, the parquet reader could drop the buffer, and fall back to decoding it again

Reply:
> If we've buffered too much in memory, which is highly likely with any non-selective predicate, the parquet reader could drop the buffer, and fall back to decoding it again

Ah, yeah that makes sense